src/docs/binding-closures.md

```diff
@@ -0,0 +1,135 @@
+
+Given a closure of the form `(arguments) -> return`, this can be converted into the form: `(UnsafeMutablePointer<return>, UnsafeMutablePointer<(arguments)>)->()` For example:
+```swift
+public func callsIntoCSharp (a: @escaping (Int, SomeStruct, Bool) -> SomeOtherStruct) {
+    let a_adapter = { (ret: UnsafeMutablePointer<SomeOtherStruct>, args: UnsafeMutablePointer<(Int, SomeStruct, Bool)>) in
+        let (i, ss, b) = args.pointee
+        ret.initialize(to: a(i, ss, b))
+    }
+    callToCSharpMethod (a_adapter)
+}
+```
+
+In this case, the original closure gets captured by `a_adapter`. Before calling the adapter, C# obviously has to create the argument tuple, space for the return value (allocated but not initialized) and can call it because effectively the delegate signature has become `delegate void csAdapter (IntPtr ret, IntPtr args)`. A similar process is used to get C# closures into Swift.
+
+One obvious problem here is that this only works with escaping closures. If rewritten with a non escaping closure, the act of capturing the original closure in the new one is flagged by the Swift compiler as an error. Fortunately, Swift has a workaround to this via the function [`withoutActuallyEscaping`](https://developer.apple.com/documentation/swift/withoutactuallyescaping(_:do:)/), for which the previous code can be rewritten as this:
+
+```swift
+public func callsIntoCSharp (a: (Int, SomeStruct, Bool) -> SomeOtherStruct) {
+    withoutActuallyEscaping (a) { a_escaping in
+        let a_adapter = { (ret: UnsafeMutablePointer<SomeOtherStruct>, args: UnsafeMutablePointer<(Int, SomeStruct, Bool)>) in
+            let (i, ss, b) = args.pointee
+            ret.initialize(to: a_escaping(i, ss, b))
+        }
+    }
+    callToCSharpMethod (a_adapter)
+}
+```
+
+Please note that with runtime support, this shouldn't be necessary, but this may be important for future reference.
+
+In running the other direction, we would need a way to convert a C# closure into something that is callable from Swift. The approach in BTfS is heavy handed because of the lack of runtime support. Given a C# closure, we create a handle to it, then call into a Swift routine which generates a swift closure that calls back into C# with a pointer to argument and return as before, but now with the handle and a `@convention (c)` function pointer to goes back into to a C# routine that unpacks the arguments, uses the handle to get the original C# closure and calls it. We should be able to do better than this.
+
+# Idiomatic Differences
+
+The main idiomatic difference has to do with the escaping/non-escaping varieties of closure. Obviously, C# doesn't make this distinction. As such, if a C# method gets passed a delegate from Swift that is non-escaping, it incumbent upon the user to never store it. We can make this somewhat better by putting an attribute on such delegates that flags it as a non-escaping and create a Roslyn analyzer that looks for usage that would violate that.
```

Comment:
Maybe non-escaping closures can be modeled using a ref struct? Ref structs ["are allocated on the stack and can't escape to the managed heap."](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/ref-struct)

Calling the closure might end up a little bit weird in C# though, since I think you'd have to call a method on the struct instead of using delegate syntax.

```cs
ref struct MyNonEscapingClosure {
    public void Invoke (...)
    {
		// call the closure here
    }
}
```